### PR TITLE
DSV4: Add shard_indexer_acts to shard the CSA indexer scores across heads

### PR DESCRIPTION
**Pull request by systalyze-ai**
# Description

Adds `shard_indexer_acts` to shard the DeepSeek-V4 CSA indexer's quadratic score tensor across `activation_heads`. This reduces peak memory at long context.

# Performance

Disabling `shard_indexer_acts` at seq 16384 exceeds the per-chip memory limit by 4.42 GiB (35.67 GiB needed, 31.25 GiB available) and the run does not compile; with it enabled the run fits. Setup: `ici_expert_parallelism=16`, `ici_tensor_parallelism=8`, `ici_fsdp_parallelism=1`, `max_target_length=16384`, `per_device_batch_size=0.125` (global batch 16), `LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=98304`, dynamic-splash attention path enabled.

# Tests

- `tests/unit/indexer_activation_sharding_test.py` — 9 passed.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by richjames0 — CHANGES_REQUESTED** [review]
I think we may want to close this one, as main has changed significantly and I **think** already addresses what this PR sought to.

```
1. The target code has been restructured out from under it

4909 inserts sharding constraints around this in DeepseekV4Indexer.__call__:

scores = jnp.einsum("bhsd,bhwd->bhsw", q, compressed_kv)
scores = jax.nn.relu(scores) * self.softmax_scale
index_scores = jnp.einsum("bhsw,bsh->bsw", scores, weights)

On main that plain einsum pair is gone. attention_compressed.py:1008-1046 is now a three-way dispatch:

if use_kernel:
  index_scores = self._streamindex_scores_sharded(q, compressed, weights)
elif head_chunk_size > 0:
  ...jax.lax.scan(jax.checkpoint(scan_body_indexer), ...)
else:
  index_scores = csa_streamindex.csa_indexer_scores_jax(q, compressed, weights, ...)

compressed_kv was also renamed to compressed and is [B, W, D] (3-D) rather than the 4-D bhwd 4909 assumes. mergeable_state: dirty confirms the conflict.

2. Main already solves the problem, twice

- _streamindex_scores_sharded (attention_compressed.py:831) wraps the StreamIndex Pallas kernel in a batch-only shard_map, explicitly to stop GSPMD all-gathering the full [B, H, S, D] query. Its docstring notes context/sequence parallelism is deliberately out of scope.
- csa_qk_head_chunk_size chunks over indexer heads with jax.checkpoint, which removes the [B, H, S, W] intermediate outright rather than sharding it.

4909's activation_heads constraint on scores is a weaker version of what the head-chunking path already achieves. Before any rewrite, it's worth establishing what 4909 still buys on top of these — I don't think that's obvious, and it should be in the PR description.
```
